framework/include/utils/InputParameterWarehouse.h

```diff
@@ -23,7 +23,32 @@
 class InputParameters;
 class MooseApp;
 
-typedef std::map<std::string, MooseSharedPointer<InputParameters> >::iterator InputParameterIterator;
+/**
+ * A storage container for the name of parameters to be controlled
+ */
+struct InputParametersContainer
+{
+  /// The "system" name (e.g., Kernels)
+  std::string system;
```

Comment:
I don't really understand why `system` needs to be separate from tags.  `tag` should be a `std::set` and should include `system/object`.